### PR TITLE
fix(agent): emit OpenAI Responses tool-call args events keyed by item_id

### LINKED ISSUE #2550: [Bug] OpenAI Responses streaming: tool-call-args-delta/tool-call-ready never emitted (callBuffers keyed by call_id, but events carry item_id) — MCPAgent.stream() yields zero steps
**Describe the bug**

On the OpenAI Responses streaming path, `streamResponsesTurn` keys its tool-call buffers (`callBuffers`) by `call_id`, but the streaming events that carry the arguments — `response.function_call_arguments.delta` and `response.function_call_arguments.done` — do not have a `call_id` field; they carry `item_id` (the output-item id, `fc_...`) and `output_index` only. So the buffer lookup key is always the fallback `""` and `callBuffers.get("")` is always `undefined`. Neither `tool-call-args-delta` nor `tool-call-ready` is ever emitted.

Consequence: `streamNativeAgentSteps` (native_runner.ts) builds AgentSteps only on `tool-call-ready`, so `MCPAgent.stream()` / `agent.streamEvents()` silently yields **zero** steps for provider `"openai"` (and openrouter models routed to `OpenAIResponsesDriver`). The tool still executes (arguments are parsed from `response.completed`'s output), so the bug is silent: the user sees the final text but no tool step at all.

**Environment**

- mcp-use agent package version: `2.0.16` (`libraries/typescript/packages/agent/package.json`)
- commit: `dcaa0b8df5721f11d4d2aefae1f85df2a1903ba5`
- Node: `v22.23.2`
- OS: macOS 26.6.2 (Darwin 25.6.0), headless Node (no browser involved)
- `openai` SDK: `6.17.0` (lockfile-matched)

**To Reproduce**

Steps:
1. Check out `dcaa0b8df5721f11d4d2aefae1f85df2a1903ba5`.
2. Run the repro below against `src/llm` with `tsx`.

The repro feeds raw OpenAI Responses SSE events with the real SDK event shapes (SSE stream served by a stubbed `globalThis.fetch`), first asserting on `driver.stream()` event types, then on `streamNativeAgentSteps`.

```ts
import { OpenAIResponsesDriver } from "../src/llm/providers/openai-responses-driver.ts";
import { streamNativeAgentSteps } from "../src/llm/native_runner.ts";
import type { LlmStreamEvent } from "../src/llm/types.ts";

// Real OpenAI Responses streaming event shapes, taken from the openai-node SDK
// (ResponseFunctionCallArgumentsDeltaEvent / ...DoneEvent carry item_id +
// output_index, NOT call_id).
const turn1 = [
  { type: "response.output_item.added", output_index: 0,
    item: { type: "function_call", id: "fc_1", call_id: "call_abc", name: "add", arguments: "" } },
  { type: "response.function_call_arguments.delta", item_id: "fc_1", output_index: 0, delta: '{"a":10' },
  { type: "response.function_call_arguments.delta", item_id: "fc_1", output_index: 0, delta: ',"b":20}' },
  { type: "response.function_call_arguments.done", item_id: "fc_1", output_index: 0, arguments: '{"a":10,"b":20}' },
  { type: "response.completed", response: { output: [
      { type: "function_call", id: "fc_1", call_id: "call_abc", name: "add", arguments: '{"a":10,"b":20}' } ] } },
];
const turn2 = [
  { type: "response.output_text.delta", delta: "The answer is 30." },
  { type: "response.completed", response: { output: [
      { type: "message", role: "assistant", content: [{ type: "output_text", text: "The answer is 30." }] } ] } },
];
const sse = (events: unknown[]) => events.map((e) => `data: ${JSON.stringify(e)}\n\n`).join("");

let call = 0;
(globalThis as any).fetch = async () => {
  call++;
  return new Response(sse(call === 1 ? turn1 : turn2), {
    status: 200,
    headers: { "Content-Type": "text/event-stream" },
  });
};

async function main() {
  const driver = new OpenAIResponsesDriver({ provider: "openai", model: "gpt-5", apiKey: "k" });
  let toolArgSeen: unknown = null;
  const events: LlmStreamEvent[] = [];
  for await (const ev of driver.stream({ messages: [{ role: "user", content: "add 10 and 20" }], tools: [] })) {
    events.push(ev);
  }
  console.log("driver.stream() event types:", JSON.stringify(events.map((e) => e.type)));
  console.log("tool-call-ready count:", events.filter((e) => e.type === "tool-call-ready").length);
  console.log("args-delta count:", events.filter((e) => e.type === "tool-call-args-delta").length);

  call = 0;
  const steps: unknown[] = [];
  const gen = streamNativeAgentSteps(driver, {
    messages: [{ role: "user", content: "add 10 and 20" }],
    tools: [],
    maxSteps: 3,
    callTool: async (_n, args) => {
      toolArgSeen = args;
      return { content: [{ type: "text", text: "30" }] };
    },
  });
  let r = await gen.next();
  while (!r.done) { steps.push(r.value); r = await gen.next(); }
  console.log("callTool received args:", JSON.stringify(toolArgSeen));
  console.log("agent.stream() steps yielded:", steps.length, JSON.stringify(steps));
  console.log("agent.stream() returned text:", JSON.stringify(r.value));
}
main();
```

**Actual output**

Verbatim from an independent verifier run (`node tsx repro2.ts`, unmodified repro above):

```
driver.stream() event types: ["tool-call-start","done"]
tool-call-ready count: 0
args-delta count: 0
callTool received args: {"a":10,"b":20}
agent.stream() steps yielded: 0 []
agent.stream() returned text: "The answer is 30."
```

A second independent verifier, running the same shape and an identical probe against the tree `src/`:

```
SHAPE A (spec: item_id only): types ["tool-call-start","done"]; args-delta 0; tool-call-ready 0
SHAPE A: callTool got {"name":"add","args":{"a":10,"b":20}}
SHAPE A: agent steps 0 []; final text "The answer is 30."
SHAPE B (control: call_id injected): args-delta 2; tool-call-ready 1
PATCHED, SHAPE A: types ["tool-call-start","tool-call-args-delta","tool-call-args-delta","tool-call-ready","done"]; args-delta 2; tool-call-ready 1
PATCHED, SHAPE A: agent steps 2; baseline vitest run src/llm 47/47 pass, patched also 47/47
```

**Expected behavior**

The OpenAI Responses path should emit `tool-call-args-delta` for each `response.function_call_arguments.delta` and one `tool-call-ready` when `response.function_call_arguments.done` arrives, matching:

- the openai-node SDK event contract — `ResponseFunctionCallArgumentsDeltaEvent` = `{ delta, item_id, output_index, sequence_number, type }` and `ResponseFunctionCallArgumentsDoneEvent` = `{ arguments, item_id, output_index, sequence_number, type }` (`openai@6.17.0`, `resources/responses/responses.d.ts`; raw.githubusercontent.com/openai/openai-node/src/resources/responses/responses.ts lines 3257 / 3287). Neither declares `call_id`. `ResponseFunctionToolCall` declares `call_id: string` *and* `id?: string` — distinct fields; `item_id` points at the output-item `id` (`fc_...`).
- the sibling provider `providers/openai-chat-completions.ts`, which emits `tool-call-args-delta` (line 205) and `tool-call-ready` (line 226).
- the internal contract `llm/types.ts:157` — `LlmToolCallReadyEvent` is documented as "A tool call with complete, parsed arguments", and `native_runner.ts:51-58` builds an AgentStep only on `tool-call-ready`.

So `agent.stream()` should yield one AgentStep for the `add` call, instead of `0 []`.

**Root cause**

`libraries/typescript/packages/agent/src/llm/providers/openai-responses.ts:328` and `:345`: both handlers read `const callId = typeof parsed.call_id === "string" ? parsed.call_id : ""`, but these events carry `item_id`, not `call_id`, so the key is always `""` and `callBuffers.get(callId)` (line 330 / line 348) is always `undefined`. The buffer is written under `item.call_id` at line 311 (populated from `response.output_item.added`, line 308). Result: the delta and done branches never emit their events, and every downstream consumer of `tool-call-ready` sees an empty step list.

Fix direction (not a full diff): key `callBuffers` by the output item id (`item.id`, i.e. `fc_...`) and look it up with `parsed.item_id` (or fall back via `output_index` to the stored index) in the delta/done handlers, while still yielding the stored `call_id` as `toolCallId` on the emitted events.

**Screenshots**

N/A — headless Node library, no UI screenshot applies.

**Desktop (please complete the following information)**

- OS: macOS 26.6.2
- Version: mcp-use agent `2.0.16` @ `dcaa0b8df5721f11d4d2aefae1f85df2a1903ba5`, Node v22.23.2, openai 6.17.0
- Browser: N/A

**Smartphone (please complete the following information)**

N/A — not a mobile issue.

**Additional context**

- Reachability: `new MCPAgent({ llm: "openai/gpt-5", ... }).stream({ prompt })` → `agents/mcp_agent.ts:549` `streamNativeAgentSteps(driver)` → `llm/native_runner.ts:52` → `createLlmDriver` returns `OpenAIResponsesDriver` (`llm/driver.ts:53-62`) → `streamToolLoop` → `streamResponsesTurn`. Same path for openrouter models matching `/^~?openai\//`. The inspector client-side chat (`useChatMessagesClientSide.ts:436`) creates the tool card at `tool-call-start` with args `{}` and never receives `tool-call-ready`.
- Related issues/PRs found by collision checks: none. Searches for `tool-call-ready`, `function_call_arguments`, and "Responses streaming" in mcp-use/mcp-use return only unrelated closed issues.
- Happy to open a PR with the approach sketched above if you'd like, or happy to be assigned.

### PR DESCRIPTION
**Pull request by BlueX888**
# Pull Request Description

## Language / Project Scope

Check all that apply:
- [x] TypeScript
- [ ] Python
- [ ] Documentation only
- [ ] CI/CD or tooling

## Changes

`OpenAIResponsesDriver` streaming never emitted `tool-call-args-delta` or
`tool-call-ready`, so `MCPAgent.stream()` / `agent.streamEvents()` yielded zero
AgentSteps for tool calls on the Responses path (provider `"openai"`, and
OpenRouter models routed to this driver). The tool still executed, so the bug
was silent.

`callBuffers` is now keyed by the output item id (`item.id`, `fc_...`), which is
what `response.function_call_arguments.delta` / `.done` actually carry, while
the emitted events keep yielding the MCP-facing `call_id`. The buffer stores
both so `tool-call-ready`'s `toolCallId` still matches the `call_id` used by
`tool-result`.

## Review Readiness

Help maintainers review this quickly:

- [x] The PR is scoped to one issue or one clearly related change
- [x] The PR description explains the user-visible problem and the fix
- [x] The diff avoids unrelated formatting, generated files, and bulk rewrites
- [x] I verified the affected package/docs path with the commands listed below
- [x] I checked for existing open PRs that already solve the same issue

## Implementation Details

1. `libraries/typescript/packages/agent/src/llm/providers/openai-responses.ts`
   - `callBuffers` is keyed by `item.id` from `response.output_item.added`
     (falling back to the `call_id` when the item has no `id`), instead of
     `call_id`.
   - The buffer value gains a `callId` field so the stored `call_id` survives.
   - `response.function_call_arguments.delta` and `...done` now look up
     `parsed.item_id`, and emit `toolCallId: buf.callId` — the `call_id`, not
     the item id.
2. No dependency changes. No public API change.
3. A patch changeset is included, as required for TypeScript changes.

### Packages Modified

Check all packages that were modified:
- [ ] `docs`
- [ ] `tests`
- [x] `cli`
- [ ] `create-mcp-use-app`
- [ ] `mcp-use` (server)
- [ ] `mcp-use` (client)
- [ ] `inspector`

`@mcp-use/agent` (the package containing `src/llm/`).

### Pre-commit Checklist

Ensure all of the following have been completed:
- [x] Ran `pnpm lint:fix` to auto-fix linting issues
- [x] Ran `pnpm format` to format code with Prettier
- [x] Ran `pnpm build` and build succeeds without errors
- [x] Ran `pnpm changeset` to create a changeset (if this PR includes user-facing changes)
- [x] Added or updated tests if needed
- [ ] Updated documentation in `docs/` folder if needed

## Python Checklist

> Complete this section if your PR includes Python changes

N/A — TypeScript only.

## Example Usage (Before)

```ts
const agent = new MCPAgent({ llm: "openai/gpt-5", mcpClient });
let steps = 0;
for await (const step of agent.stream({ prompt: "What is the weather in Paris?" })) {
  steps++;
}
// steps === 0 — the final text arrives, but the tool call produces no step
```

## Example Usage (After)

```ts
const agent = new MCPAgent({ llm: "openai/gpt-5", mcpClient });
let steps = 0;
for await (const step of agent.stream({ prompt: "What is the weather in Paris?" })) {
  steps++;
}
// steps === 2 — one step on tool-call-ready, one on tool-result
```

## Documentation Updates

* None — internal streaming event mapping, no documented behavior changes.

## Testing

**Regression test** (`libraries/typescript/packages/agent/src/llm/__tests__/openai-responses.test.ts`),
using the real Responses SSE event shapes with `fetch` stubbed:

- `emits tool-call-args-delta and tool-call-ready against the call_id` — asserts
  the driver's event stream contains the two `tool-call-args-delta` events and
  the `tool-call-ready` with `toolCallId: "call_abc"`.
- `yields AgentSteps for a streamed OpenAI Responses tool call` — runs the full
  `streamNativeAgentSteps(OpenAIResponsesDriver)` path and asserts the two
  AgentSteps.

Both fail on the unpatched source and pass after the fix. The pre-existing
`parses function_call_arguments.done into tool-call-ready shape` case was
replaced: it only asserted `JSON.parse` of an inline object literal and never
called into `streamResponsesTurn`, and it encoded the incorrect assumption that
the done event carries `call_id`.

```bash
cd libraries/typescript/packages/agent

# new test — 8 passed (8)
pnpm vitest run src/llm/__tests__/openai-responses.test.ts

# module tests — 10 files, 48 passed (48)
pnpm vitest run src/llm

# full package unit suite — 18 files, 140 passed | 1 skipped
pnpm test:unit

# typecheck — no output, exit 0
pnpm exec tsc --noEmit

cd libraries/typescript

# format — All matched files use Prettier code style!
pnpm prettier --check packages/agent/src/llm/providers/openai-responses.ts \
  packages/agent/src/llm/__tests__/openai-responses.test.ts

# lint — no output, exit 0
pnpm eslint packages/agent/src/llm/providers/openai-responses.ts \
  packages/agent/src/llm/__tests__/openai-responses.test.ts

# changeset — @mcp-use/agent patch
pnpm changeset status --since=main
```

Edge cases considered: an `output_item.added` item without `id` falls back to the
existing `call_id`/`call_{n}` key (unchanged behavior for that case), and an
unknown `item_id` in a delta/done still yields nothing rather than throwing.

## Backwards Compatibility

Backwards compatible. Consumers that only read the final text are unaffected;
consumers of `tool-call-ready` / `MCPAgent.stream()` now receive the steps that
were previously dropped. The `toolCallId` on emitted events is still the
`call_id`, so tool-result correlation is unchanged.

## Related Issues

Closes #2550

---
## Summary by cubic
Fixes OpenAI Responses streaming so tool calls now emit `tool-call-args-delta` and `tool-call-ready` events. Previously, the driver looked up argument events by `call_id` even though Responses sends `item_id`, so `MCPAgent.stream()` and `streamEvents()` dropped tool-call AgentSteps; the tool still executed, so the bug was silent.

**Bug Fixes**

- Buffers argument events under both the output item id and the provider `call_id`, using `item:`/`call:`-prefixed keys so one call's item id can't collide with another call's `call_id`.
- Emitted events keep the provider `call_id` in `toolCallId`, so tool-result correlation is unchanged.
- Adds regression coverage for both SSE correlation paths, an identifier-collision case, and end-to-end AgentSteps.

<sup>Written for commit 24174d0e2b95e70805317b7347d2097a359a3268. Summary will update on new commits.</sup>

<a href="https://cubic.dev/pr/mcp-use/mcp-use/pull/2551?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

### CODE REVIEW

**Review by MohammedAlkindi — COMMENTED** [review]
The fix is correct for the real API. `response.function_call_arguments.delta` and `.done` carry `item_id`, while the buffer was keyed by the `call_id` read off `response.output_item.added`, so `callBuffers.get(callId)` never matched and neither event was ever emitted.

One narrowing worth a decision. The old lookup read `parsed.call_id`, so a producer that identifies argument events by `call_id` matched before and does not now, because the buffer is keyed on `item.id` whenever that field is present. #2553 was opened about five hours after this one, makes the same `item_id` change, and keeps a fallback that also matches a buffer by its stored `callId`. The two PRs disagree only on that compatibility path, so it is probably worth picking one rather than carrying both.

The `streamNativeAgentSteps` test is the stronger half here. It pins the symptom a user actually reports, AgentSteps never arriving for a tool call, rather than only the event shape.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by BlueX888** [comment]
Thanks — both points taken, and the narrowing is gone in 519e7c23.

On picking one: @Elioooon withdrew #2553 the same morning, so this is the only PR left on #2550 and there is nothing to reconcile. That also removed the reason to keep the two behaviours mutually exclusive, so I kept the compatibility path rather than dropping it.

The buffer is now indexed under both identities. The item id is tried first and the fallback only runs when it misses, so a spec-conformant producer takes exactly the path you verified against the real API:

```ts
// response.output_item.added
callBuffers.set(typeof item.id === "string" ? item.id : callId, buffer);
callBuffers.set(callId, buffer);

// response.function_call_arguments.{delta,done}
const buf = callBuffers.get(itemId) ?? callBuffers.get(callId);
```

Emitted events still report the protocol `call_id` as `toolCallId`, unchanged.

Verification, run in an installed workspace (`pnpm --filter @mcp-use/agent test:run src/llm/__tests__/openai-responses.test.ts`):

- 9 passed. `toolCallEvents` now takes a `correlation` argument, and the new case feeds the same sequence with the argument events carrying `call_id`.
- I checked the new test is load-bearing rather than decorative: on the previous commit it fails with `expected [...(2)] to deep equally contain { type: 'tool-call-args-delta', ... }` — only `tool-call-start` comes through — and it passes here. The existing `item_id` test passes on both.
- `prettier --check` clean on both files.
- The full `@mcp-use/agent` suite is 12 failed / 64 passed on this head, and 12 failed / 62 passed on a clean `main` in the same workspace; the failures are all `Failed to resolve entry for package "@mcp-use/client"` from an unbuilt workspace dependency, so they are not from this change. Hosted CI stays authoritative for that reason.

File: libraries/typescript/packages/agent/src/llm/providers/openai-responses.ts [hunk 1]
```diff
@@ -308,12 +314,18 @@ export async function* streamResponsesTurn(
...
+        const buffer = {
           index: idx,
+          callId,
           name,
           argsJson: "",
           started: true,
-        });
+        };
+        callBuffers.set(typeof item.id === "string" ? item.id : callId, buffer);
+        callBuffers.set(callId, buffer);
```
[the remark this answers is not in the document]
Comment (BlueX888) [reply libraries/typescript/packages/agent/src/llm/providers/openai-responses.ts]: Correct, and reproduced before fixing it — this one is mine. `24174d0e` prefixes the keys.

A two-call stream where the second call's `call_id` equals the first call's item id showed the failure directly at `519e7c23`. The first call's argument events were emitted against the second call:

```
{ type: 'tool-call-args-delta', index: 1, toolCallId: 'fc_shared', toolName: 'second_tool' }
{ type: 'tool-call-ready',     index: 1, toolCallId: 'fc_shared', toolName: 'second_tool' }
```

The second `output_item.added` stored its buffer under `fc_shared` because that is its `call_id`, overwriting the entry the first call had stored under `fc_shared` because that is its item id. The argument events name the first call's item id, so they resolved to the second call's buffer. The two identifiers come from the producer and share no namespace, so they are now stored as `item:<id>` and `call:<callId>` and cannot overlap. That test is part of the commit.

File: libraries/typescript/packages/agent/src/llm/__tests__/openai-responses.test.ts [hunk 2]
```diff
@@ -146,15 +149,211 @@ describe("responsesReasoningFields", () => {
...
+    expect(events).toContainEqual({
+      type: "tool-call-ready",
+      index: 0,
+      toolCallId: "call_abc",
+      toolName: "get_weather",
+      args: { city: "Paris" },
+    });
+  });
+
+  it("still correlates argument events from a call_id-only producer", async () => {
```
[the remark this answers is not in the document]
Comment (BlueX888) [reply libraries/typescript/packages/agent/src/llm/__tests__/openai-responses.test.ts]: Right — the test name claimed more than the fixture did. `24174d0e` parameterizes it over both producer shapes: one that sets the output item's `id` and one that omits it, so the `typeof item.id === "string" ? item.id : callId` fallback is exercised rather than only the second key.

[the remark this answers is not in the document]
Comment (BlueX888) [reply libraries/typescript/packages/agent/src/llm/__tests__/openai-responses.test.ts]: Taken as fixing the duplication, though not by folding the two into a loop over `["item_id", "call_id"]`: the two call_id-only producer shapes now share one `it.each` body, which removes the duplication and adds the missing coverage at the same time. The item_id test stays separate because it exercises the documented wire shape rather than a producer variation.
